### ISSUE TITLE
Use non-blocking sockets for I/O

### ISSUE THREAD

**Issue by alanxz** [alanxz/rabbitmq-c#227]
Implement non-blocking sockets for I/O. This will help with detecting heatbeat timeouts on `amqp_basic_publish()` or anything that does a `send()`.

This should be portable and usable on most platforms (unlike setting timeouts on `recv()` or `send()` operations which supposedly doesn't work at all on Win32).

I suspect this will require quite a bit of refactoring to avoid turning the socket code into more of a mess. Its possible with enough re-factoring we'll end up with something closer to the way [c-ares](http://c-ares.haxx.se/) works, though I believe we're a long way from that.

**Comment by alanxz** [comment]
This will likely fix #175

**Comment by goughy** [comment]
Is there any update on this issue?  I work for a small software house and we use this (excellent) library on embedded Windows CE devices (don't ask!).  This issue is biting us in the field because of really bad 3G comms and we get hung comms on a daily basis.

Is there any way we can bring forward this feature?  I thought about adding the feature myself, but I think that may suffer poor implementation quality :-), although I may have to try depending on timings.

PM me if you like

**Comment by alanxz** [comment]
This is something I'm actively working on, I'd estimate I'll have something worth trying out in 2-4 weeks. If this is something you're interested in helping out with, let me know and I can give you some pointers on how to get started.

**Comment by goughy** [comment]
That would be awesome, and I'd be happy to try.  Let me know how to get started, and what process you'd like to adopt (fork & PR I assume?).  Is there a current branch with anything on it I can look at?

The current version we have is 0.3.0, so the code has moved on a bit since then.  I see in the codebase now that both receiving and connecting are non-blocking, so its just the actual send that requires wrapping in 'select()'?  

If that happens, is your plan that the socket can just stay non-blocking the whole time and blocking semantics can be simulated with null timeouts in select()?

**Comment by alanxz** [comment]
I accept changes through PRs, so your best bet is to fork this repo, do your development work, then send a PR once you have something that is ready for review.

Currently the library does not use non-blocking sockets outside of the [`amqp_open_socket_noblock()`](https://github.com/alanxz/rabbitmq-c/blob/bfd8cc9129da46136691a88c06d505b78f0043ae/librabbitmq/amqp_socket.c#L267), function. Instead the library tries to determine whether the socket is readable/writable by calling `poll()` before attempting to read or write from the socket.  [Aside: the library switched from `select()` to `poll()` as the former has a socket-id range issue on some platforms, Linux and MacOSX specifically, `select()` and `poll()` functionally do very similar things].

What needs to change is the sockets that are handled in [`amqp_tcp_socket.c`](https://github.com/alanxz/rabbitmq-c/blob/d7029dbc2c3088cf5bf3c5044ef2fee7a1b12dcc/librabbitmq/amqp_tcp_socket.c) need to be put in non-blocking mode, then instead of doing `poll()` before trying to read or write from the socket, the read or write call should be done first, then if a would-block error is returned, use `poll()` to wait as necessary for the socket to become readable or writeable.

There are two things that will make this somewhat complex:
- The code that handles heartbeats is somewhat intertwined with this socket timeout code, and its a bit messy, some refactoring will be required here.
- OpenSSL can handle non-blocking sockets, but it has its own requirements for how to handle non-blocking sockets. SSL support will be required before this can be merged into master.

I would recommend starting with making the non-SSL stuff work first before trying to tackle the SSL case.

**Comment by goughy** [comment]
I've made a start, but there are a few design questions I would like to ask, and a quick review before I go any further.

Could you have a look at:
https://github.com/goughy/rabbitmq-c/commit/154147d3d5487d3702f71b2d847eef151c44e51d

Maybe there is an easier way to communicate?  Email/IM etc  Preference?

Initial questions:
1. Can we always leave the the socket in non-blocking mode and handle blocking semantics as a NULL timeout?  It should work, but requires handling in close() also.
2. Do you envisage timeouts set against the socket (as in the commit), or passed through from the higher level API's? ie either the client calls amqp_tcp_socket_set_send_timeout() and all internal calls to the socket class send() or writev() will honour the timeout, or is it passed through the various layers on each call?

**Comment by alanxz** [comment]
Lets have the conversation here, that way there's a record of the design decisions made.

1. Yes. For blocking behavior the flow should look like `read()` [not enough bytes], `poll()` for socket readability with a -1 timeout, and loop on that until enough bytes are read.

2. The timeout needs to be a higher-level abstraction.  E.g., it should not be read-with-timeout, its more of a try to read from this socket, then have a different function to test the socket for readabilty/writeability. There are two reasons for this: there is a situation where we need to wait for socket-readability, socket-writeability, or a timeout (heartbeats are enabled, and are publishing to a connection that is in flow-control).  The OpenSSL has a similar issue, doing a non-blocking read(), can have the library return a WANT_WRITE, in the case of a key renegotiation for example.
